prov/gni/src/gnix_msg.c

```diff
@@ -262,40 +332,104 @@ int _gnix_ep_eager_msg_w_data_match(struct gnix_fid_ep *ep, void *msg,
 			gnix_slist_insert_tail(&req->slist,
 					       &ep->pending_recv_comp_queue);
 		fastlock_release(&ep->recv_comp_lock);
-	} else {
 
-		req = _gnix_fr_alloc(ep);
-		if (req == NULL) {
-			ret = -FI_ENOMEM;
-			goto err;
-		}
+	} else {
 
-		req->loc_addr = (uint64_t)malloc(len);
-		if (req->loc_addr == 0UL) {
+		req->loc_addr = (uint64_t)malloc(hdr->len);
+		if (unlikely(req->loc_addr == 0UL)) {
 			_gnix_fr_free(ep, req);
```

Comment:
the req can be freed here since basically we've run out of memory trying to handle unexpected short messages.  I'll add a comment here saying we'll eventually need better error recover, flow control, or something.